pkg/generate/crds/gen.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crds
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	extensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/gengo/args"
+	"sigs.k8s.io/controller-tools/pkg/generate"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen"
+	"sigs.k8s.io/controller-tools/pkg/internal/codegen/parse"
+)
+
+type CRDGenerator struct {
+	RootPath          string
+	Namespace         string
+	SkipMapValidation bool
+}
+
+func (c *CRDGenerator) GenerateCRDs() error {
+	arguments := args.Default()
+	b, err := arguments.NewBuilder()
+	if err != nil {
+		return fmt.Errorf("failed making a parser: %v", err)
+	}
+
+	if len(c.RootPath) > 0 && !pathHasProjectFile(c.RootPath) {
+		return fmt.Errorf("input path must be a diretory containing %s", "PROJECT")
+	}
+
+	// Search for PROJECT if no rootPath provided
+	if len(c.RootPath) == 0 {
+		c.RootPath, err = filepath.Abs(".")
+		if err != nil {
+			return err
+		}
+
+		for {
+			if pathHasProjectFile(c.RootPath) {
+				break
+			}
+
+			if isGoPath(c.RootPath) {
+				return fmt.Errorf("failed to find working directory that contains %s", "PROJECT")
+			}
+
+			c.RootPath = path.Dir(c.RootPath)
+		}
+	}
+
+	// Switch working directory to root path.
+	if err := os.Chdir(c.RootPath); err != nil {
+		return fmt.Errorf("failed switching working dir: %v", err)
+	}
+
+	if err := b.AddDirRecursive("./pkg/apis"); err != nil {
+		return fmt.Errorf("failed making a parser: %v", err)
+	}
+	ctx, err := parse.NewContext(b)
+	if err != nil {
+		return fmt.Errorf("failed making a context: %v", err)
+	}
+
+	arguments.CustomArgs = &parse.ParseOptions{SkipMapValidation: c.SkipMapValidation}
+
+	// TODO: find an elegant way to fulfill the domain.
+	p := parse.NewAPIs(ctx, arguments, getDomainFromProject())
```

Comment:
Lets make `domain` also an input to generator and remove the dependency from the `PROJECT` file completely.